lib_src/api/asynchronous_fifo.h

```diff
@@ -0,0 +1,208 @@
+ *
+ * @param   Ki                  Integral constant for the FIFO.
+ *                              This gets multiplied by the phase error
+ *                              measured in ticks (typically -20,000 - 20,000)
+ *                              and added to the ratio_error.
+ *                              A typical value is 200 - 300.
+ *
+ * @param  ticks_between_samples The number of ticks between samples is used to
+ *                              estimate the expected phase error halfway down
+ *                              the FIFO.
+ */
+void asynchronous_fifo_init_PID_raw(asynchronous_fifo_t * UNSAFE state,
+                                    int Kp, int Ki, int ticks_between_samples);
+
+/**
+ * Function that that resets the FIFO from the producer side. Either this function should
+ * be called on the producing side, or ``asynchronous_fifo_reset_consumer()``
+ * should be called on the consumer side. In both cases the whole FIFO will be reset back
+ *
+ * @param   state               Asynchronous FIFO to be initialised
+ */
+void asynchronous_fifo_reset_producer(asynchronous_fifo_t * UNSAFE state);
+
+/**
+ * Function that that resets the FIFO from the consumer side. Either this function should
+ * be called on the consuming side, or ``asynchronous_fifo_reset_producer()``
+ * should be called on the producer side. In both cases the whole FIFO will be reset back
+ *
+ * @param   state               Asynchronous FIFO to be initialised
+ */
+void asynchronous_fifo_reset_consumer(asynchronous_fifo_t * UNSAFE state);
+
+/**
+ * Function that must be called to deinitalise the asynchronous FIFO
+ *
+ * @param   state                   ASRC structure to be de-initialised
+ */
+void asynchronous_fifo_exit(asynchronous_fifo_t * UNSAFE state);
+
+/**
+ * Function that provides the next sample to the asynchronous FIFO.
+ *
+ * This function and the consume function both need a timestamp,
+ * which is the time that the last sample was input (this function) or
+ * output (``asynchronous_fifo_consume``). The asynchronous FIFO will hand the
+ * samples across from producer to consumer through an elastic queue, and run
+ * a PID algorithm to calculate the best way to equalise the input clock relative
+ * to the output clock. Therefore, the timestamps
+ * have to be measured on either the same clock or two very similar clocks.
+ * It is probably fine to use the reference clocks on two tiles, provided
+ * the tiles came out of reset at more or less the same time. Using the
+ * clocks from two different chips would require the two chips to share an
+ * oscillator, and for them to come out of reset simultaneously.
+ *
+ * @param   state               ASRC structure to push the sample into
+ *
+ * @param   samples             The sample values.
+ *
+ * @param   n                   The number of samples
+ *
+ * @param   timestamp           The number of ticks when this sample was input.
+ *
```

Comment:
xscope_used param missing. I think it would be good to remove this from the API. Defines can be messy but does feel like a good place for a define (debug mode effectively)